### PR TITLE
Convert HexSet to be a HexMap

### PR DESCRIPTION
**Pull request by JayKickliter**
This PR is a major change. It turns HexSet into `HexMap<Value>`. Think of it as a dictionary of Hex → Value, but one that understands the hierarchical nature of H3. You can query a small hex for a value, but possibly get returned the value for a 'complete' parent hex. If all you wanted was a Hex → Value dictionary, a regular `HashMap<H3Cell, V>` would do.

## What about sets?

With this change, what we `HexSet` is represented as a `HexMap<()>`, where we're using the unit (`()`) type for values. Unfortunately, this does not optimize into the equivalent of the previous `HexSet` type and imposes fairly large performance penalties. The performance hits are in the neighborhood of 130% on my 2013 Mac Pro. But keep in mind that means that calls to `.contains()` now take ~40 ns when they used to take ~17 ns. 40 ns is still pretty damn fast. 

EDIT: I switched to criterion for proper benchmarking and the performance difference is negligible 

## Remaining work

- [x] decide if pluggable compactor is part of the map or only provided with `.insert_and_compact()`
- [ ] ability to reduce a hex?
- [x] update documentation to reflect this major change
- [ ] ~make HexSet its own type to avoid `Value = ()` perf hit?~ out of scope
- [x] rename this repo `hexmap` right before/after merging

The first point about where the compactor should live has major implications for serialization, the namability of types, etc. For example, what if the user wants to use a closure for compaction? That's currently [supported](). But if the compactor was a generic member of the HexMap type itself, how do you name it? Hint: you can't name the type of closure.

On the second point about reduction: what if, instead of getting the value for a Hex, the user wants to perform a reduction/fold on all children for a particular hex? Reduction could be very useful when you want to say, summarize all data for all hexes in a region. That would be hard to do without exposing the internal `Node` type.

### CODE REVIEW

**Comment by Vagabond** [comment]
Maybe we should add this as a second type in the library so other uses don't need to pay the perf penalty?

**Comment by JayKickliter** [comment]
I suspect the new overhead can be optimized away when using the unit type for Value. In theory, it should be possible to have no penalty at all when the value is zero sized.

File: src/hexset.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
+//! A HexMap is a structure for representing geographical regions and
```
Comment (maplant) [inline src/hexset.rs:1]: Because this module is private you should move these docs into the the root module

**Comment by maplant** [comment]
Do you expect users will call `insert_and_compact` with different compactors? If not, it might make sense to define the compactor as a type argument to `HexMap` and remove the option to `insert` without compact, similarly to how `HashMap` works: 

```rust
struct HexMap<V, C = EqCompactor> { ... }
```

**Comment by JayKickliter** [comment]
> Do you expect users will call insert_and_compact with different compactors? If not, it might make sense to define the compactor as a type argument to HexMap and remove the option to insert without compact

I do not, but can rule out anyone will. And at the same time, compaction can only happen when walking up the stack after an insert/update. But, my original idea is exactly what you suggest. The main downside to making the compactor part of the HexMap itself is serialization and naming.  Serailization is simple to solve, but possibly lossy on the deserialization side. The bigger problem is that it's impossible to name `struct HexMap<V, C>` when C is a closure.

**Comment by maplant** [comment]
So, to your point about naming closures for the compactor - this is actually possible in some sense by creating a helper function for creation: 

```rust
impl HexMap<V,  C> { 
    pub fn new_with_compactor(compactor: C) -> Self { ... }
}
```

And that will be sufficient for using this struct with a closure compactor outside of structs.

As for spelling this out in structs, there are several ways. If it's a closure, you need to do:

```rust
struct X {
    map: HexMap<V, Box<dyn FnMut(&V) -> bool>>,
}
```

if it's a function, then you can do `&'static dyn FnMut(&V) -> bool`

**Comment by JayKickliter** [comment]
@maplant thanks for the ideas. I think you're right on all counts. I'm not sure why I'm so obsessed with the closure problem, other than an annoyance I had with another crate recently.

**Pull request merged**